File: app/src/main/java/wangdaye/com/geometricweather/common/basic/models/options/appearance/HourlyTrendDisplay.java
package wangdaye.com.geometricweather.common.basic.models.options.appearance;

import android.content.Context;
import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.StringRes;

import java.util.ArrayList;
import java.util.List;

import wangdaye.com.geometricweather.R;

public enum HourlyTrendDisplay {

    TAG_TEMPERATURE("temperature", R.string.temperature),
    TAG_WIND("wind", R.string.wind),
    TAG_PRECIPITATION("precipitation", R.string.precipitation);

    private final String value;
    private @StringRes final int nameId;

    HourlyTrendDisplay(String value, int nameId) {
        this.value = value;
        this.nameId = nameId;
    }

    public String getTagValue() {
        return value;
    }

    public String getTagName(Context context) {
        return context.getString(nameId);
    }

    @NonNull
    public static List<HourlyTrendDisplay> toHourlyTrendDisplayList(String value) {
        if (TextUtils.isEmpty(value)) {
            return new ArrayList<>();
        }
        try {
            String[] cards = value.split("&");

            List<HourlyTrendDisplay> list = new ArrayList<>();
            for (String card : cards) {
                switch (card) {
                    case "temperature":
                        list.add(TAG_TEMPERATURE);
                        break;

                    case "wind":
                        list.add(TAG_WIND);
                        break;

                    case "precipitation":
                        list.add(TAG_PRECIPITATION);
                        break;
                }
            }
            return list;
        } catch (Exception e) {
            return new ArrayList<>();
        }
    }

    @NonNull
    public static String toValue(@NonNull List<HourlyTrendDisplay> list) {
        StringBuilder builder = new StringBuilder();
        for (HourlyTrendDisplay v : list) {
            builder.append("&").append(v.getTagValue());
        }
        if (builder.length() > 0 && builder.charAt(0) == '&') {
            builder.deleteCharAt(0);
        }
        return builder.toString();
    }

    @NonNull
    public static String getSummary(Context context, @NonNull List<HourlyTrendDisplay> list) {
        StringBuilder builder = new StringBuilder();
        for (HourlyTrendDisplay v : list) {
            builder.append(",").append(v.getTagName(context));
        }
        if (builder.length() > 0 && builder.charAt(0) == ',') {
            builder.deleteCharAt(0);
        }
        return builder.toString().replace(",", ", ");
    }
}

File: domain/src/main/java/wangdaye/com/geometricweather/common/basic/models/options/appearance/HourlyTrendDisplay.kt
package wangdaye.com.geometricweather.common.basic.models.options.appearance

import android.content.Context
import androidx.annotation.StringRes
import wangdaye.com.geometricweather.core.R
import wangdaye.com.geometricweather.common.basic.models.options._basic.BaseEnum

enum class HourlyTrendDisplay(
    override val id: String,
    @StringRes val nameId: Int
): BaseEnum {

    TAG_TEMPERATURE("temperature", R.string.temperature),
    TAG_WIND("wind", R.string.wind),
    TAG_UV_INDEX("uv_index", R.string.uv_index),
    TAG_PRECIPITATION("precipitation", R.string.precipitation);

    companion object {

        @JvmStatic
        fun toHourlyTrendDisplayList(
            value: String
        ) = if (value.isEmpty()) {
            emptyList()
        } else try {
            value.split("&").mapNotNull { card ->
                when (card) {
                    "temperature" -> TAG_TEMPERATURE
                    "wind" -> TAG_WIND
                    "uv_index" -> TAG_UV_INDEX
                    "precipitation" -> TAG_PRECIPITATION
                    else -> null
                }
            }
        } catch (e: Exception) {
            emptyList()
        }

        @JvmStatic
        fun toValue(list: List<HourlyTrendDisplay>): String =
            list.joinToString("&") { it.id }

        @JvmStatic
        fun getSummary(context: Context, list: List<HourlyTrendDisplay>): String =
            list.joinToString(", ") { it.getName(context) }
    }

    override val valueArrayId = 0
    override val nameArrayId = 0

    override fun getName(context: Context) = context.getString(nameId)
}
